Refactor contract, oracle, chain

davidyuk opened the pull request.

thepiwo commented on es/ae/oracle.js:
```diff
@@ -61,27 +62,23 @@ async function getOracleObject (oracleId) {
+  const fetchQueries = async () => {
+    const queries = ((await this.getOracleQueries(oracleId)).oracleQueries || [])
+      .filter(({ id }) => !sentQueryIds.has(id))
+    queries.forEach(({ id }) => sentQueryIds.add(id))
```
I'd refactor this to use map again.

(not tested)
```javascript
const allQueries = (await this.getOracleQueries(oracleId)).oracleQueries || [])
const queries = [...new Set(allQueries.map(({ id }) => id))]
```

davidyuk replied:
It looks like we don't need to store all previous queries, just the last one instead, and pass it as `from` parameter to GetOracleQueriesByPubkey. I would like to check it later, proposing to leave as it is now.
https://github.com/aeternity/aepp-sdk-js/issues/1055

thepiwo commented on es/ae/oracle.js:
```diff
@@ -61,27 +62,23 @@ async function getOracleObject (oracleId) {
-    }
+function pollForQueries (oracleId, onQuery, { interval = 5000 } = {}) {
+  const sentQueryIds = new Set()
+  const fetchQueries = async () => {
```
fetch implies to me something is returned, but we just want the `onQuery` to be triggered

davidyuk replied:
I'm proposing a better naming in 5164358a931215ef629608c6d1369f302d966c75

thepiwo commented on es/ae/oracle.js:
```diff
@@ -121,22 +118,15 @@ async function getQueryObject (oracleId, queryId) {
-    if (left > 0) {
-      await pause(interval)
-      return probe.bind(this)(left - 1)
+  for (let i = 0; i < attempts; i++) {
```
are you sure the for correctly awaits the promise here for each iteration?

I'd maybe go for a more high-level approach.

1. use a range array instead of `for i` (e.g. as in https://github.com/aeternity/aepp-governance/blob/master/governance-server/src/util.js#L45)
2. use an async reduce to iterate and eventually break the iteration (maybe something similar to https://github.com/aeternity/aepp-governance/blob/master/governance-server/src/util.js#L58 can be helpful)

thepiwo replied:
This could also be combined in a util function, e.g. as `attemptWithPause(() => ...)` passing the function, where e.g. a rejected promise or returned false is the break condition

davidyuk replied:
> are you sure the for correctly awaits the promise here for each iteration?

Yes, I think so. Actually, I'm rarely writing a `for` cycle, but I think it is a good place to use it. I prefer this because it uses just plain js.

thepiwo commented on es/chain/node.js:
```diff
@@ -115,54 +108,35 @@ async function height () {
-    throw Error(`Giving up after ${attempts * interval}ms, current=${current}, h=${h}`)
+async function awaitHeight (height, { interval = 5000, attempts = 20 } = {}) {
+  let currentHeight
+  for (let i = 0; i < attempts; i++) {
```
same as above applies

thepiwo commented:
I didn't notice the minor breaking changes, maybe a summary of what the PR does with a list of breaking changes would be helpful in the PR description.

davidyuk commented:
I have rebased this PR on develop and fixed tests.

> I didn't notice the minor breaking changes

The only one was a making pollForQueries to run in sync, I have made it return a dumb promise to be able to merge it earlier.

davidyuk commented:
I have just restarted tests and they have passed 🤷‍♀️
